File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def bfs(maps):
    queue = deque([[0, 0]])
    dx = [+1, -1, 0, 0]
    dy = [0, 0, +1, -1]
    
    while queue:
        row_idx, col_idx = queue.popleft()
        for i in range(4):
            new_row = row_idx + dx[i]
            new_col = col_idx + dy[i]
            if 0 <= new_row < len(maps) and 0 <= new_col < len(maps[0]) and maps[new_row][new_col] == 1:
                queue.append([new_row, new_col])
                maps[new_row][new_col] = maps[row_idx][col_idx] + 1
                
    return maps[len(maps) - 1][len(maps[0]) - 1] if maps[len(maps) - 1][len(maps[0]) - 1] > 1 else -1
        

def solution(maps):
    return bfs(maps)

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js


function solution(maps) {
    const queue = [[0, 0, 1]];
    const visited = [];
    const n = maps.length;
    const m = maps[0].length;
    for (let i = 0; i < n; i++) {
        const temp = [];
        for (let j = 0; j < m; j++) {
            temp.push(false);
        }
        visited.push(temp);
    }
    
    function BFS() {
        while(queue.length > 0) {
            const size = queue.length;
            for (let i = 0; i < size; i++) {
                const [x, y, count] = queue.shift();
                if(x === n - 1 && y === m - 1) {
                    return count;
                }
                const dx = [1, -1, 0, 0];
                const dy = [0, 0, 1, -1];

                for (let i = 0; i < 4; i++) {
                    const nx = x + dx[i];
                    const ny = y + dy[i];
                    if(nx >= 0 && ny >= 0 && nx < n && ny < m && maps[nx][ny] === 1 && visited[nx][ny] === false) {
                        visited[nx][ny] = true;
                        queue.push([nx, ny, count + 1]);
                    }
                }
                
            }
        }
        return -1;
    }
    return BFS()
}